### ISSUE TITLE
Map a pyQuirk circuit to  an Interlin-q circuit.

### ISSUE THREAD

**Issue by stephendiadamo** [Interlin-q/Interlin-q#35]
In Interlin-q, one has to create potentially complicated circuits. It could be easier to have a graphical interface to create them. We can, as a first step, integrate pyQuirk with Interlin-q for circuit creation:

- https://github.com/adgt/pyQuirk

**Comment by Ritvik-Sapra** [comment]
Hello, could you please provide some more detail and nudge me in the right direction?

**Comment by stephendiadamo** [comment]
@Ritvik-Sapra. Sorry for the delay. The goal of this task is to be able to generate circuits for Interlin-q using pyQuirk. Essentially what we want to do here is build a circuit using pyQuirk and take the circuit components to map them to the circuit for Interlin-q. Interlin-q also uses circuits, one builds them at the moment purely from code. We want to simplify this process by having a user interface. 

The task is essentially: 
- Build up the circuit from pyQuirk
- Extract the circuit model in the pyQuirk framework
- Use that mapping to build an Interlin-q circuit 
- Use that circuit to the Interlin-q simulation by giving the circuit to the controller host

We can discuss more but as a first step of extracting the pyQuirk circuit would be the point of discussion for the next steps.

**Comment by Ritvik-Sapra** [comment]
Please have a look at [https://github.com/abhayhegde/pyquirk](this repo: https://github.com/abhayhegde/pyquirk)
Here we can simply use the URL or JSON file generated by the quirk, and generate an output file in .tex format. 
Can this be helpful?

**Comment by stephendiadamo** [comment]
Yes, the goal would be to, all within the Python framework, extract the JSON object from PyQuirk and then use it to build the Interlin-q circuit. From the JSON object you can iterate through the objects and programmatically build a Interlin-q circuit.

**Comment by theRoughCode** [comment]
Hey @Ritvik-Sapra, is this being worked on?

**Comment by Ritvik-Sapra** [comment]
> Hey @Ritvik-Sapra, is this being worked on?

Hey, I am stuck in some unavoidable work right now. Feel free to take it forward.

**Comment by zemenfes-afk** [comment]
I am stuck with this problem

Exception in thread Thread-25:
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.1520.0_x64__qbz5n2kfra8p0\lib\threading.py", line 954, in _bootstrap_inner
    self.run()

**Comment by stephendiadamo** [comment]
@zemenfes-afk I've removed your comment as it's not related to this issue.

**Comment by zemenfes-afk** [comment]
oky where is the comment area to post.

**Comment by stephendiadamo** [comment]
@zemenfes-afk Best to create a new issue with a bug tag.

**Comment by theRoughCode** [comment]
@Ritvik-Sapra cool, thanks.

@stephendiadamo to clarify, is this meant to be used only in python notebooks?

**Comment by stephendiadamo** [comment]
As a starting point that would be fine I'd say. Maybe one could have a command line script that pops up a circuit builder at some point, but I think in the notebook is enough for this step.

**Comment by theRoughCode** [comment]
I have a few questions regarding the specifications of `Operation`. Are you available on the UF discord for DMs?

**Comment by stephendiadamo** [comment]
Sure, feel free to write me. I'm in the quantum networks channel

**Comment by zemenfes-afk** [comment]
i have little questions on the interlin-q program,I tryto run it but not
working?

**Comment by stephendiadamo** [comment]
@zemenfes-afk Since this is not related to the issue you've posted on I'm going to remove your comment. Can you open up a new issue with your bugs? Also it is not possible for me to help you without knowing what you tried, your OS, your environment, etc. Could you include these details?

**Comment by theRoughCode** [comment]
We didn't get to release this in time for UnitaryHack, but here are some notes from our discussion so far for the next person to tackle/answer:

When mapping from pyQuirk, it's not obvious how to determine which computing host to assign to each qubit. Thus, a way around this is to have the mapping function take in a list of ordered qubits with their respective computing hosts:
```python

qubits = [
  Qubit(computing_host_id="QPU_0", q_id="q_0_0"),
  Qubit(computing_host_id="QPU_0", q_id="q_0_1"),
  Qubit(computing_host_id="QPU_1", q_id="q_1_0"),
  Qubit(computing_host_id="QPU_1", q_id="q_1_1")
]

def from_pyquirk(json: dict, qubits : List[Qubit], q_map: dict) -> Circuit:
  ...
```

The way I envision it might work is as follows:
1. Caller passes in pyQuirk JSON and ordered list of N qubits together with the `q_map` topology.
2. Convert each gate in the JSON into its corresponding `Operation` and apply it to the relevant `Qubit`.
3. Combine the gates into `Layer`s and create the corresponding `Circuit` using `q_map`.
4. Return the `Circuit`.

Other things to think about:
1. Quirk allows specifying the input qubit to be 1 instead of 0. You would need to add an X gate or augment the `PREPARE_QUBITS` constant to handle initializing to a 1 state.
2. SWAP gates and controlled-Z gates aren't provided by the package, so you would have to implement them. @stephendiadamo mentioned that there might be some issues with Qutip backend from QuNetSim when implementing custom gates.

**Comment by anushkrishnav** [comment]
Is this issue available to be worked on ?
or is anyone else working on it ?
I read over the older conversations getting a rough idea but a task list would be really helpful as in whats needed can get on a call via discord for more discussions

**Comment by stephendiadamo** [comment]
No I don't believe so. Do you have some specific questions I can answer asynchronously? Might not be easy to find a time for a call in the next few days.
